cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,348 @@
+    auto right_key_table = device_table::create(right_key_cols_vect.size(), right_key_cols_vect.data());
+    rmm::device_vector<int8_t> asc_desc_d(asc_desc);
+
+    rmm::device_vector<index_type> merged_indices = generate_merged_indices(*left_key_table, *right_key_table, asc_desc_d, nulls_are_smallest, stream);
+
+    // Allocate output table
+    bool nullable = has_nulls(left_sync_table) || has_nulls(right_sync_table);
+    table destination_table(left_sync_table.num_rows() + right_sync_table.num_rows(), column_dtypes(left_sync_table), nullable, false, stream);
+    for (gdf_size_type i = 0; i < destination_table.num_columns(); i++) {
+        gdf_column const* left_col = left_sync_table.get_column(i);
+        gdf_column * out_col = destination_table.get_column(i);
+        
+        if (left_col->dtype != GDF_STRING_CATEGORY){
+            continue;
+        }
+
+        NVCategory * category = static_cast<NVCategory*>(left_col->dtype_info.category);
+        out_col->dtype_info.category = category->copy();
+    }
+    
+    // Materialize
+    auto left_device_table_ptr = device_table::create(left_sync_table, stream);
+    auto right_device_table_ptr = device_table::create(right_sync_table, stream);
+    auto output_device_table_ptr = device_table::create(destination_table, stream);
+    auto& left_device_table = *left_device_table_ptr;
+    auto& right_device_table = *right_device_table_ptr;
+    auto& output_device_table = *output_device_table_ptr;
+
+    auto index_start_it = thrust::make_zip_iterator(thrust::make_tuple(
+                                                    thrust::make_counting_iterator(static_cast<gdf_size_type>(0)), 
+                                                    merged_indices.begin()));
+    auto index_end_it = thrust::make_zip_iterator(thrust::make_tuple(
+                                                thrust::make_counting_iterator(static_cast<gdf_size_type>(merged_indices.size())),
+                                                merged_indices.end()));
+
+    thrust::for_each(rmm::exec_policy(stream)->on(stream),
+                    index_start_it,
+                    index_end_it,
+                    [=] __device__ (auto const & idx_tuple){
+                        gdf_size_type dest_row = thrust::get<0>(idx_tuple);
+                        index_type merged_idx = thrust::get<1>(idx_tuple);
+                        side src_side = thrust::get<0>(merged_idx);
+                        gdf_size_type src_row  = thrust::get<1>(merged_idx);
+                        device_table const & src_device_table = src_side == side::LEFT ? left_device_table : right_device_table;
+                        copy_row<false>(output_device_table, dest_row, src_device_table, src_row);
+                    });
+    
+    CHECK_STREAM(0);
+
+    if (nullable) {
+        for (gdf_size_type i = 0; i < destination_table.num_columns(); i++) {
```

Comment:
This could be nicely implemented as a `std::transform` where the inputs are `left_sync_table` and `right_sync_table`, and the output is `destination_table`. `table::begin` and `table::end` are provided for exactly this kind of thing.

Reply:
I'm not following. Do you mean something like this?
```c++
std::transform(left_sync_table.begin(),
               left_sync_table.end(),
               right_sync_table.begin(),
               destination_table.begin(),
               [&, =stream](gdf_column const* left_col, gdf_column const* right_col){
                   /* destination_table already has the merged data, 
                    * need the same gdf_column to update the valids
                    * and null_count and return it */
                   int index = /*calculate which column index*/;
                   gdf_column* out_col = destination_table.get_column(index);
                   materialize_bitmask(left_col, right_col, out_col, merged_indices.data().get(), stream);
                   out_col->null_count = left_col->null_count + right_col->null_count;
                   return out_col;
               });
```

Reply:
Ah, I see the problem is you need to modify `out_col` in place. Nevermind.